genesyscloud/routing_queue/genesyscloud_routing_queue_proxy.go

```diff
@@ -0,0 +1,163 @@
+	RoutingQueueCache                rc.CacheInterface[platformclientv2.Queue]
+}
+
+// newRoutingQueuesProxy initializes the routing queue proxy with all the data needed to communicate with Genesys Cloud
+func newRoutingQueuesProxy(clientConfig *platformclientv2.Configuration) *RoutingQueueProxy {
+	api := platformclientv2.NewRoutingApiWithConfig(clientConfig)
+	routingQueueCache := rc.NewResourceCache[platformclientv2.Queue]()
+
+	return &RoutingQueueProxy{
+		clientConfig:                     clientConfig,
+		routingApi:                       api,
+		getAllRoutingQueuesAttr:          getAllRoutingQueuesFn,
+		getRoutingQueueByIdAttr:          getRoutingQueueByIdFn,
+		getRoutingQueueWrapupCodeIdsAttr: getRoutingQueueWrapupCodeIdsFn,
+		RoutingQueueCache:                routingQueueCache,
+	}
+}
+
+// GetRoutingQueueProxy acts as a singleton to for the internalProxy.  It also ensures
+// that we can still proxy our tests by directly setting internalProxy package variable
+func GetRoutingQueueProxy(clientConfig *platformclientv2.Configuration) *RoutingQueueProxy {
+	if internalProxy == nil {
+		internalProxy = newRoutingQueuesProxy(clientConfig)
+	}
+	return internalProxy
+}
+
+// GetAllRoutingQueues retrieves all Genesys Cloud routing queues
+func (p *RoutingQueueProxy) GetAllRoutingQueues(ctx context.Context) (*[]platformclientv2.Queue, *platformclientv2.APIResponse, error) {
+	return p.getAllRoutingQueuesAttr(ctx, p)
+}
+
+// getRoutingQueueById returns a single Genesys Cloud Routing Queue by ID
+func (p *RoutingQueueProxy) getRoutingQueueById(ctx context.Context, queueId string) (*platformclientv2.Queue, *platformclientv2.APIResponse, error) {
+	return p.getRoutingQueueByIdAttr(ctx, p, queueId)
+}
+
+// getRoutingQueueWrapupCodeIds returns a list of routing queue wrapup code ids
+func (p *RoutingQueueProxy) getRoutingQueueWrapupCodeIds(ctx context.Context, queueId string) ([]string, *platformclientv2.APIResponse, error) {
+	return p.getRoutingQueueWrapupCodeIdsAttr(ctx, p, queueId)
+}
+
+// getAllRoutingQueuesFn is the implementation for retrieving all routing queues in Genesys Cloud
+func getAllRoutingQueuesFn(ctx context.Context, p *RoutingQueueProxy) (*[]platformclientv2.Queue, *platformclientv2.APIResponse, error) {
+	var allQueues []platformclientv2.Queue
+	const pageSize = 100
+
+	queues, resp, getErr := p.routingApi.GetRoutingQueues(1, pageSize, "", "", nil, nil, nil, false)
+	if getErr != nil {
+		return nil, resp, fmt.Errorf("failed to get first page of queues: %v", getErr)
+	}
+
+	// Check if the routing queue cache is populated with all the data, if it is, return that instead
+	// If the size of the cache is the same as the total number of queues, the cache is up-to-date
+	if rc.GetCacheSize(p.RoutingQueueCache) == *queues.Total {
+		return rc.GetCache(p.RoutingQueueCache), nil, nil
+	} else if rc.GetCacheSize(p.RoutingQueueCache) != *queues.Total && rc.GetCacheSize(p.RoutingQueueCache) != 0 {
```

Comment:
Should this not be done after collecting all the other pages? This is just the first page of queues

Reply:
No I put it here intentionally, my logic is when I do the first api call we can use the `Total` value that is returned to check if the number of routing queues in the org is equal to the number of routing_queues in the cache. If it is then I assume the cache is up to date with all the data and we don't need to do the rest of the api calls adn can use the cache instead

Reply:
Ah right of course. Missed the total part, assumed you were checking length